plugin/src/rules/prefer_destructuring.rs

```diff
@@ -0,0 +1,950 @@
+use std::sync::Arc;
+
+use once_cell::sync::Lazy;
+use serde::Deserialize;
+use squalid::{EverythingExt, OptionExt};
+use tree_sitter_lint::{
+    rule, tree_sitter::Node, violation, Fixer, NodeExt, QueryMatchContext, Rule,
+};
+
+use crate::{
+    ast_helpers::{get_number_literal_value, NodeExtJs, Number, NumberOrBigInt},
+    kind,
+    kind::{
+        AssignmentExpression, Identifier, Kind, MemberExpression, PrivatePropertyIdentifier,
+        PropertyIdentifier, SubscriptExpression, Super, VariableDeclarator,
+    },
+    utils::{ast_utils, ast_utils::get_static_string_value},
+};
+
+#[derive(Copy, Clone, Debug, Default, Deserialize)]
+#[serde(default)]
+struct ArrayAndObject {
+    array: bool,
+    object: bool,
+}
+
+#[derive(Copy, Clone, Debug, Default, Deserialize)]
+#[serde(default, rename_all = "PascalCase", deny_unknown_fields)]
+struct ByNodeType {
+    variable_declarator: Option<ArrayAndObject>,
+    assignment_expression: Option<ArrayAndObject>,
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(untagged)]
+enum FirstOption {
+    ByNodeType(ByNodeType),
+    ArrayAndObject(ArrayAndObject),
+}
+
+impl FirstOption {
+    fn normalized(&self) -> ByNodeType {
+        match self {
+            Self::ByNodeType(value) => *value,
+            Self::ArrayAndObject(value) => ByNodeType {
+                variable_declarator: Some(*value),
+                assignment_expression: Some(*value),
+            },
+        }
+    }
+}
+
+#[derive(Debug, Default, Deserialize)]
+#[serde(default)]
+struct SecondOption {
+    enforce_for_renamed_properties: bool,
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(untagged)]
+enum Options {
+    Single([FirstOption; 1]),
```

Comment:
Sort of tried to also support supply just this "first" option without surrounding `[...]` but didn't seem super easy to get that to work (seemed like it was then ignoring the second one) so whatever might be "nice to have" if I understood what the `Deserialize` pattern would have to be to support that